pact/sale-contracts/dutch-auction/dutch-auction.pact

```diff
@@ -0,0 +1,210 @@
+    reserve-price:decimal
+    start-price:decimal)
+    (enforce (> start-date (curr-time)) "Start date must be in the future")
+    (enforce (> end-date start-date) "End date must be after start date")
+    (enforce (> reserve-price 0.0) "Reserve price must be greater than 0")
+    (enforce (> start-price reserve-price) "Start price must be greater than reserve price")
+  )
+
+  (defun create-auction
+    ( sale-id:string
+      token-id:string
+      start-date:integer
+      end-date:integer
+      reserve-price:decimal
+      start-price:decimal    
+    )
+    (validate-auction start-date end-date reserve-price start-price)
+    (let (
+      (quote-info:object{quote-schema} (get-quote-info sale-id)))
+
+      (enforce (= (at 'sale-price quote-info) 0.0) "Quote price must be 0")
+      (enforce (= (at 'sale-type quote-info) "marmalade-sale.dutch-auction") "Quote does not support auction")
+      (enforce (= token-id (at 'token-id quote-info)) "Token-id does not match quote token-id")
+    )
+
+    (with-capability (MANAGE_AUCTION sale-id token-id)
+      (insert auctions sale-id {
+        "token-id": token-id
+        ,"start-date": start-date
+        ,"end-date": end-date
+        ,"start-price": start-price
+        ,"reserve-price": reserve-price
+        ,"sell-price": 0.0
+        ,"buyer": ""
+        ,"buyer-guard": DUMMY_GUARD
+      })
+      (emit-event (AUCTION_CREATED sale-id token-id))
+    )
+  )
+
+  (defun update-auction
+    ( sale-id:string
+      start-date:integer
+      end-date:integer
+      start-price:decimal
+      reserve-price:decimal
+    )
+    (validate-auction start-date end-date reserve-price start-price)
+
+    (with-read auctions sale-id
+      { 'token-id:= token-id,
+        'start-date:= curr-start-date
+      }
+      (enforce (> curr-start-date (curr-time)) "Can't update auction after it has started")
+
+      (with-capability (MANAGE_AUCTION sale-id token-id)
+        (update auctions sale-id {
+          "start-date": start-date
+          ,"end-date": end-date
+          ,"start-price": start-price
+          ,"reserve-price": reserve-price
+        })
+      )
+    )
+  )
+
+  (defun retrieve-auction (sale-id:string)
+    (read auctions sale-id)
+  )
+
+  (defun get-current-price:decimal (sale-id:string)
+    @doc "Returns the current price of the auction"
+    (with-read auctions sale-id
+      {
+        'start-date:= start-date,
+        'end-date:= end-date,
+        'reserve-price:= reserve-price,
+        'start-price:= start-price
+      }
+
+      (let* (
+        (sale-period-hours:decimal  (* (round (/ (- end-date start-date) 3600.0)) 1.0))
+        (period-passed-hours:decimal (* (round (/ (- (curr-time) start-date) 3600.0)) 1.0))
+        (price-range:decimal (- start-price reserve-price))
+        )
+        (if (or (< (curr-time) start-date) (> (curr-time) end-date))
+          0.0
+          (round (- start-price (* (/ period-passed-hours sale-period-hours) price-range)) 2)
```

Comment:
The interval is already set, the price now drops every hour so that is predictable. 

I am looking into making the interval configurable per auction as a result of some offline feedback from @wooglie